pkg/apis/seedmanagement/types_managedseed.go

```diff
@@ -0,0 +1,161 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package seedmanagement
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	gardencore "github.com/gardener/gardener/pkg/apis/core"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ManagedSeed represents a Shoot that is registered as Seed.
+type ManagedSeed struct {
+	metav1.TypeMeta
+	// Standard object metadata.
+	metav1.ObjectMeta
+	// Specification of the ManagedSeed.
+	Spec ManagedSeedSpec
+	// Most recently observed status of the ManagedSeed.
+	Status ManagedSeedStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ManagedSeedList is a list of ManagedSeed objects.
+type ManagedSeedList struct {
+	metav1.TypeMeta
+	// Standard list object metadata.
+	metav1.ListMeta
+	// Items is the list of ManagedSeeds.
+	Items []ManagedSeed
+}
+
+// ManagedSeedSpec is the specification of a ManagedSeed.
+type ManagedSeedSpec struct {
+	// Shoot references a Shoot that should be registered as Seed.
+	Shoot Shoot
+	// SeedTemplate is a template for a Seed object, that should be used to register a given cluster as a Seed.
+	// Either SeedTemplate or Gardenlet must be specified. When Seed is specified, the ManagedSeed controller will not deploy a gardenlet into the cluster
+	// and an existing gardenlet reconciling the new Seed is required.
+	SeedTemplate *SeedTemplate
+	// Gardenlet specifies that the ManagedSeed controller should deploy a gardenlet into the cluster
+	// with the given deployment parameters and GardenletConfiguration.
+	Gardenlet *Gardenlet
+}
+
+// Shoot identifies the Shoot that should be registered as Seed.
+type Shoot struct {
+	// Name is the name of the Shoot that will be registered as Seed.
+	Name string
+}
+
+// SeedTemplate is a template for creating a Seed object, when registering a cluster as a ManagedSeed.
+type SeedTemplate struct {
+	// Standard object metadata.
+	metav1.ObjectMeta
+	// Specification of the desired behavior of the Seed.
+	Spec gardencore.SeedSpec
+}
+
+// Gardenlet specifies gardenlet deployment parameters and the GardenletConfiguration used to configure gardenlet.
+type Gardenlet struct {
+	// Deployment specifies certain gardenlet deployment parameters, such as the number of replicas,
+	// the image, etc.
+	Deployment *GardenletDeployment
+	// Config is the GardenletConfiguration used to configure gardenlet.
+	Config *runtime.RawExtension
```

Comment:
`Config` should be of type `runtime.Object` in the internal API version, see https://github.com/kubernetes/apimachinery/blob/0e9f9cff521d6ebb5aaee87ae496b501bc27e974/pkg/runtime/types.go#L49-L94
```suggestion
	Config runtime.Object
```

Reply:
I tried this earlier but couldn't make it work. Will give it another try. Perhaps you are aware that `RawExtension`, and not `Object` is used in internal versions of all other Gardener APIs that use this type.

Reply:
Yes, but let's do it the right way this time then, as it is supposed to be done by apimachinery.

Reply:
OK, figured it out finally, will do as suggested. To make sure that it works, I had to rebase all downstream PRs up to the last one in order to test the actual controller reconciliation ... several times, until I got it right. This was **not** fun. I hope you are indeed having an easier time with the multiple PRs, because it's definitely not the case on my side.